pkg/ccl/changefeedccl/alter_changefeed_stmt.go

```diff
@@ -305,15 +167,443 @@ func alterChangefeedPlanHook(
+// with each target we are adding/removing (i.e. with initial_scan or
+// no_initial_scan), and the current status of the job. If the progress does not
+// need to be updated, we will simply return the previous progress and statement
+// time that is passed into the function
+func generateNewProgress(
+	prevProgress jobspb.Progress,
+	prevStatementTime hlc.Timestamp,
+	existingTargetSpans []roachpb.Span,
+	newSpans []roachpb.Span,
+	withInitialScan bool,
+) (jobspb.Progress, hlc.Timestamp, error) {
+	prevHighWater := prevProgress.GetHighWater()
+	changefeedProgress := prevProgress.GetChangefeed()
+
+	var spanGroup roachpb.SpanGroup
+
+	if changefeedProgress != nil && changefeedProgress.Checkpoint != nil {
+		spanGroup.Add(changefeedProgress.Checkpoint.Spans...)
+	}
+	spanGroup.Add(newSpans...)
+
+	// if the high watermark is empty
+	if prevHighWater == nil || prevHighWater.IsEmpty() {
+		// if we want to perform an initial scan on the new targets
+		if withInitialScan {
+			// do not make any changes to the progress
+			return prevProgress, prevStatementTime, nil
+		}
+
+		// if we do not want to perform an initial scan on these targets, update the
+		// checkpoint to include the spans of these targets so that the changefeed
+		// will skip performing a backfill on these targets
+		newProgress := jobspb.Progress{
+			Progress: &jobspb.Progress_HighWater{},
+			Details: &jobspb.Progress_Changefeed{
+				Changefeed: &jobspb.ChangefeedProgress{
+					Checkpoint: &jobspb.ChangefeedProgress_Checkpoint{
+						Spans: spanGroup.Slice(),
+					},
+				},
+			},
+		}
+		return newProgress, prevStatementTime, nil
+	}
+
+	// if the checkpoint is non-empty
+	if changefeedProgress != nil && changefeedProgress.Checkpoint != nil && len(changefeedProgress.Checkpoint.Spans) != 0 {
+		// if we want to perform an initial scan on the new targets
+		if withInitialScan {
+			return prevProgress, prevStatementTime, errors.Errorf(`cannot perform initial scan on newly added targets during a non-initial backfill, please wait for the backfill to complete before adding these targets.`)
+		}
+
+		// if we do not want to perform an initial scan on these targets, update the
+		// checkpoint to include the spans of these targets so that the changefeed
+		// will skip performing a backfill on these targets
+		newProgress := jobspb.Progress{
+			Progress: &jobspb.Progress_HighWater{},
+			Details: &jobspb.Progress_Changefeed{
+				Changefeed: &jobspb.ChangefeedProgress{
+					Checkpoint: &jobspb.ChangefeedProgress_Checkpoint{
+						Spans: spanGroup.Slice(),
+					},
+				},
+			},
+		}
+		return newProgress, prevStatementTime, nil
+	}
+
+	// if we do not want to perform an initial scan on the new targets, and the
+	// high watermark is non-empty but the checkpoint is empty, we do not need to
+	// update the progress of the changefeed since there are no backfills in
+	// progress already
+	if !withInitialScan {
+		return prevProgress, prevStatementTime, nil
```

Comment:
Does this mean that the "Statement Time" of a changefeed now means "Either when you initially ran `CREATE CHANGEFEED`, or the last time you ran `ALTER CHANGEFEED ADD ... WITH initial_scan`, but not if you ran `ALTER CHANGEFEED` without an initial scan"

Reply:
yeah that is right. As discussed offline, we're going to rename the StatementTime to ScanTime so that the field name stays consistent with its meaning.